app/src/androidTest/java/org/mozilla/focus/helper/PrivateSessionLoadedIdlingResource.java

```diff
@@ -0,0 +1,245 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.helper;
+
+import android.graphics.Bitmap;
+
+import androidx.test.espresso.IdlingResource;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+import mozilla.components.browser.session.Download;
+import mozilla.components.browser.session.Session;
+import mozilla.components.browser.session.SessionManager;
+import mozilla.components.browser.session.manifest.WebAppManifest;
+import mozilla.components.browser.session.tab.CustomTabConfig;
+import mozilla.components.concept.engine.HitResult;
+import mozilla.components.concept.engine.media.Media;
+import mozilla.components.concept.engine.permission.PermissionRequest;
+import mozilla.components.concept.engine.prompt.PromptRequest;
+import mozilla.components.concept.engine.window.WindowRequest;
+
+
+/**
+ * An IdlingResource implementation that waits until the BrowserFragment is not loading anymore.
+ * Only after loading has completed further actions will be performed.
+ */
+public class PrivateSessionLoadedIdlingResource implements IdlingResource {
```

Comment:
Sadly we don't have any UI test (only screengrab) using this IdlingResource.
I'd suggest adding some real UI tests and make good use of it.